server/map_tree_updater.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"context"
+	"sync"
+
+	"github.com/google/trillian"
+	"github.com/google/trillian/extension"
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/merkle/smt"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/tree"
+	stree "github.com/google/trillian/storage/tree"
+	"golang.org/x/sync/errgroup"
+)
+
+type mapTreeUpdater struct {
+	tree   *trillian.Tree
+	hasher hashers.MapHasher
+	opts   TrillianMapServerOptions
+	reg    extension.Registry
+}
+
+// update updates the sparse Merkle tree at the write revision of tx with the
+// given leaf changes, and writes it to the storage. Returns the new root hash.
+func (t *mapTreeUpdater) update(ctx context.Context, tx storage.MapTreeTX, upd []smt.NodeUpdate) ([]byte, error) {
+	readRev, err := tx.ReadRevision(ctx)
+	if err != nil {
+		return nil, err
+	}
+	writeRev, err := tx.WriteRevision(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO(pavelkalinnikov): Return an error when no updates are given.
```

Comment:
It is a desired feature to be able to advance the revision number with no changes. 
IE, this updates the signed map root timestamp

Reply:
Right, but I wonder whether this deserves a different endpoint. Root signing seems like an orthogonal thing to tree building. It feels artificial to reuse `SetLeaves` RPC for that purpose.

Reply:
The API is staying as-is for the moment. We're working on improving performance within the bounds of the current API. If this causes performance problems then we can consider changing it, but otherwise we'll revisit this when we consider what shape the Map API should take when we move this from experimental.